File: iosApp/iosApp/Features/Home/MoonWheel.swift
import SwiftUI

/// Circular cycle ring — SwiftUI port of the web app's MoonWheel SVG.
/// Phase arcs (active one thicker), predicted ovulation/period markers, the
/// current-day dot, and a serif "Day N" center.
struct MoonWheel: View {
    let cycleDay: Int
    let cycleLength: Int
    let phaseLabel: String            // "Menstrual" / "Follicular" / "Ovulatory" / "Luteal"
    var size: CGFloat = 280

    private var segments: [(from: Double, to: Double, name: String, color: Color)] {
        let half = Double(cycleLength / 2)
        let len = Double(cycleLength)
        return [
            (0,        5,        "Menstrual",  Theme.phaseMenstrual),
            (5,        half - 2, "Follicular", Theme.phaseFollicular),
            (half - 2, half + 2, "Ovulatory",  Theme.phaseOvulatory),
            (half + 2, len,      "Luteal",     Theme.phaseLuteal),
        ]
    }

    private var activeColor: Color {
        segments.first { $0.name == phaseLabel }?.color ?? Theme.primary
    }

    var body: some View {
        ZStack {
            Canvas { ctx, sz in
                let r = sz.width / 2 - 20
                let c = CGPoint(x: sz.width / 2, y: sz.height / 2)

                // Center glow in the active phase colour.
                let glowR = r - 30
                ctx.fill(
                    Path(ellipseIn: CGRect(x: c.x - glowR, y: c.y - glowR, width: glowR * 2, height: glowR * 2)),
                    with: .radialGradient(
                        Gradient(colors: [activeColor.opacity(0.28), activeColor.opacity(0.06), activeColor.opacity(0)]),
                        center: c, startRadius: 0, endRadius: glowR
                    )
                )

                // Outer track.
                ctx.stroke(
                    Path(ellipseIn: CGRect(x: c.x - r, y: c.y - r, width: r * 2, height: r * 2)),
                    with: .color(Theme.inkSoft.opacity(0.22)), lineWidth: 1
                )

                // Phase arcs (active thicker + opaque).
                for s in segments {
                    let active = s.name == phaseLabel
                    ctx.stroke(
                        arc(from: s.from, to: s.to, center: c, radius: r),
                        with: .color(s.color.opacity(active ? 1 : 0.35)),
                        style: StrokeStyle(lineWidth: active ? 8 : 4, lineCap: .round)
                    )
                }

                // Predicted ovulation + period markers.
                marker(ctx, day: Double(cycleLength / 2), center: c, radius: r, color: Theme.phaseOvulatory)
                marker(ctx, day: Double(cycleLength),     center: c, radius: r, color: Theme.phaseMenstrual)

                // Current-day dot (accent) with a soft outer ring.
                let p = point(((Double(cycleDay) - 1) / Double(cycleLength)) * 360 - 90, center: c, radius: r)
                ctx.stroke(Path(ellipseIn: CGRect(x: p.x - 14, y: p.y - 14, width: 28, height: 28)),
                           with: .color(Theme.accent.opacity(0.5)), lineWidth: 1)
                ctx.fill(Path(ellipseIn: CGRect(x: p.x - 10.5, y: p.y - 10.5, width: 21, height: 21)),
                         with: .color(Theme.background))
                ctx.fill(Path(ellipseIn: CGRect(x: p.x - 9, y: p.y - 9, width: 18, height: 18)),
                         with: .color(Theme.accent))
            }

            VStack(spacing: 6) {
                Text("Day \(cycleDay)")
                    .font(.system(size: 48, weight: .regular, design: .serif))
                    .foregroundStyle(Theme.ink)
                Text(phaseLabel.uppercased())
                    .font(.caption).tracking(1.5)
                    .foregroundStyle(Theme.inkSoft)
            }
        }
        .frame(width: size, height: size)
        // Canvas is invisible to VoiceOver — expose one meaningful element.
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("Cycle day \(cycleDay) of \(cycleLength), \(phaseLabel) phase")
    }

    private func point(_ deg: Double, center: CGPoint, radius: CGFloat) -> CGPoint {
        let rad = deg * .pi / 180
        return CGPoint(x: center.x + radius * CGFloat(cos(rad)), y: center.y + radius * CGFloat(sin(rad)))
    }

    private func arc(from: Double, to: Double, center: CGPoint, radius: CGFloat) -> Path {
        var path = Path()
        let steps = max(2, Int((to - from) * 4))
        for i in 0...steps {
            let day = from + (to - from) * Double(i) / Double(steps)
            let pt = point((day / Double(cycleLength)) * 360 - 90, center: center, radius: radius)
            if i == 0 { path.move(to: pt) } else { path.addLine(to: pt) }
        }
        return path
    }

    private func marker(_ ctx: GraphicsContext, day: Double, center: CGPoint, radius: CGFloat, color: Color) {
        let p = point(((day - 1) / Double(cycleLength)) * 360 - 90, center: center, radius: radius)
        let ring = CGRect(x: p.x - 6, y: p.y - 6, width: 12, height: 12)
        ctx.fill(Path(ellipseIn: ring), with: .color(Theme.background))
        ctx.stroke(Path(ellipseIn: ring), with: .color(color), lineWidth: 2)
        ctx.fill(Path(ellipseIn: CGRect(x: p.x - 2.5, y: p.y - 2.5, width: 5, height: 5)), with: .color(color))
    }
}

File: androidApp/src/main/kotlin/app/cycluna/android/features/home/MoonWheel.kt
package app.cycluna.android.features.home

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.cycluna.android.designsystem.Theme
import app.cycluna.android.designsystem.serif
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.sin

private data class Segment(val from: Double, val to: Double, val name: String, val color: Color)

/**
 * Circular cycle ring — phase arcs with the active one thicker, predicted ovulation and
 * period markers, the current-day dot, and a serif "Day N" centre.
 */
@Composable
fun MoonWheel(
    cycleDay: Int,
    cycleLength: Int,
    periodLength: Int,
    phaseLabel: String,
    modifier: Modifier = Modifier,
    size: Dp = 260.dp,
) {
    // Same boundaries as the core's phase model (`Cycle.status`) — the arc under the
    // current-day dot must be the phase the label names, whatever the period length.
    // Menstrual wins overlaps there too, so later phases start no earlier than it ends;
    // a squeezed-out phase becomes a zero-width segment, which the drawing loop skips.
    val half = (cycleLength / 2).toDouble()
    val len = cycleLength.toDouble()
    val period = periodLength.toDouble()
    val folEnd = max(period, half - 2)
    val ovEnd = max(folEnd, half + 2)
    val segments = listOf(
        Segment(0.0, period, "Menstrual", Theme.phaseMenstrual),
        Segment(period, folEnd, "Follicular", Theme.phaseFollicular),
        Segment(folEnd, ovEnd, "Ovulatory", Theme.phaseOvulatory),
        Segment(ovEnd, len, "Luteal", Theme.phaseLuteal),
    )
    val activeColor = segments.firstOrNull { it.name == phaseLabel }?.color ?: Theme.primary

    Box(
        modifier
            .size(size)
            // A Canvas is invisible to TalkBack — expose one meaningful element instead.
            .semantics {
                contentDescription = "Cycle day $cycleDay of $cycleLength, $phaseLabel phase"
            },
        contentAlignment = Alignment.Center,
    ) {
        Canvas(Modifier.size(size)) {
            val r = this.size.width / 2f - 20.dp.toPx()
            val c = Offset(this.size.width / 2f, this.size.height / 2f)

            // Centre glow in the active phase colour.
            val glowR = r - 30.dp.toPx()
            drawCircle(
                brush = Brush.radialGradient(
                    colors = listOf(
                        activeColor.copy(alpha = 0.28f),
                        activeColor.copy(alpha = 0.06f),
                        activeColor.copy(alpha = 0f),
                    ),
                    center = c,
                    radius = glowR,
                ),
                radius = glowR,
                center = c,
            )

            // Outer track.
            drawCircle(Theme.inkSoft.copy(alpha = 0.22f), r, c, style = Stroke(1.dp.toPx()))

            // Phase arcs. Zero-width segments are skipped — a degenerate arc with a round
            // cap renders as a stray dot.
            segments.filter { it.to > it.from }.forEach { s ->
                val active = s.name == phaseLabel
                drawPath(
                    path = arcPath(s.from, s.to, c, r, cycleLength),
                    color = s.color.copy(alpha = if (active) 1f else 0.35f),
                    style = Stroke(
                        width = (if (active) 8 else 4).dp.toPx(),
                        cap = StrokeCap.Round,
                    ),
                )
            }

            // Predicted ovulation + period markers.
            drawMarker((cycleLength / 2).toDouble(), c, r, cycleLength, Theme.phaseOvulatory)
            drawMarker(cycleLength.toDouble(), c, r, cycleLength, Theme.phaseMenstrual)

            // Current-day dot, with a soft outer ring.
            val p = pointOnRing(((cycleDay - 1).toDouble() / len) * 360.0 - 90.0, c, r)
            drawCircle(Theme.accent.copy(alpha = 0.5f), 14.dp.toPx(), p, style = Stroke(1.dp.toPx()))
            drawCircle(Theme.background, 10.5.dp.toPx(), p)
            drawCircle(Theme.accent, 9.dp.toPx(), p)
        }

        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            Text("Day $cycleDay", style = serif(48).copy(color = Theme.ink))
            Text(
                phaseLabel.uppercase(),
                fontSize = 12.sp,
                letterSpacing = 1.5.sp,
                color = Theme.inkSoft,
            )
        }
    }
}

private fun pointOnRing(degrees: Double, center: Offset, radius: Float): Offset {
    val rad = degrees * Math.PI / 180.0
    return Offset(
        center.x + radius * cos(rad).toFloat(),
        center.y + radius * sin(rad).toFloat(),
    )
}

/** The arc as a polyline — dense enough that the curve reads as smooth at any ring size. */
private fun arcPath(from: Double, to: Double, center: Offset, radius: Float, cycleLength: Int): Path {
    val steps = max(2, ((to - from) * 4).toInt())
    return Path().apply {
        for (i in 0..steps) {
            val day = from + (to - from) * i / steps
            val pt = pointOnRing((day / cycleLength) * 360.0 - 90.0, center, radius)
            if (i == 0) moveTo(pt.x, pt.y) else lineTo(pt.x, pt.y)
        }
    }
}

private fun DrawScope.drawMarker(day: Double, center: Offset, radius: Float, cycleLength: Int, color: Color) {
    val p = pointOnRing(((day - 1) / cycleLength) * 360.0 - 90.0, center, radius)
    drawCircle(Theme.background, 6.dp.toPx(), p)
    drawCircle(color, 6.dp.toPx(), p, style = Stroke(2.dp.toPx()))
    drawCircle(color, 2.5.dp.toPx(), p)
}
